File: native/windows/DirectPairingWindow.cs
using System.Text.Json.Nodes;

namespace WorkspaceObservatory;

// This window owns the setup helper and the collector reservation together.
internal sealed class DirectPairingWindow : Form
{
    private readonly TlsSetupProcess bridge;
    private readonly Action<bool> release;
    private readonly System.Windows.Forms.Timer poll = new() { Interval = 2000 };
    private readonly FlowLayoutPanel content = new() { Dock = DockStyle.Fill, FlowDirection = FlowDirection.TopDown, WrapContents = false, AutoScroll = true, Padding = new Padding(18) };
    private readonly TextBox address = new(), peerAddress = new(), port = new() { Text = "43128" }, peerPort = new() { Text = "43128" };
    private readonly TextBox invitation = new() { MaxLength = 2048 };
    private readonly CheckBox consent = new() { Text = "Allow a device identity stored in a permission-restricted file", AutoSize = true };
    private readonly CheckBox ubuntu = new() { Text = "Include Ubuntu Codex records on this PC", AutoSize = true };
    private readonly Label status = new() { AutoSize = true, MaximumSize = new Size(540, 0) };
    private readonly Label fingerprint = new() { AutoSize = true, MaximumSize = new Size(540, 0) };
    private readonly List<Button> actions = [];
    private string? pendingFingerprint;
    private bool busy, hosting, joining, polling, closing, closed;
    private int generation;
    private Task? shutdown;

    internal DirectPairingWindow(TlsSetupProcess bridge, Action<bool> release)
    {
        this.bridge = bridge; this.release = release;
        Text = "Direct device pairing"; ClientSize = new Size(620, 720); MinimumSize = new Size(580, 640);
        StartPosition = FormStartPosition.CenterScreen;
        Controls.Add(content);
        Note("Choose Host on one device and Join on the other. Use a local network or trusted VPN. Provider sign-in and allowance sharing remain separate.");
        content.Controls.Add(consent);
        Note("The key is not stored using DPAPI or encrypted separately by Observatory. A valid existing identity is reused.");
        Field("This PC's numeric private IP", address); Field("This PC's sync port", port);
        Field("Mac numeric private IP", peerAddress); Field("Mac sync port", peerPort);
        content.Controls.Add(ubuntu);
        ActionButton("Create invitation", Host);
        ActionButton("Confirm this device", ConfirmHost);
        Field("Private invitation", invitation);
        ActionButton("Copy invitation", () =>
        {
            if (hosting && invitation.Text.Length > 0) Clipboard.SetText(invitation.Text);
            return Task.CompletedTask;
        });
        Note("Share invitations privately. They expire within ten minutes. Clipboard managers may retain copied invitations.");
        ActionButton("Join with invitation", Join);
        ActionButton("Finish joining", () => joining ? Send(new() { ["action"] = "join-confirm", ["includeUbuntu"] = ubuntu.Checked }) : Task.CompletedTask);
        content.Controls.Add(fingerprint); content.Controls.Add(status);
        var cancel = new Button { Text = "Cancel setup", AutoSize = true };
        cancel.Click += async (_, _) =>
        {
            if (closing) return;
            var token = ++generation; busy = true; EnableActions();
            try { var reply = await bridge.Send(new() { ["action"] = "cancel" }); if (Current(token)) ApplyReply(reply); }
            catch { if (Current(token)) Failed(); }
            finally { if (Current(token)) { busy = false; EnableActions(); } }
        };
        content.Controls.Add(cancel);
        poll.Tick += async (_, _) =>
        {
            if (closing || busy || polling || !hosting) return;
            polling = true; var token = generation;
            try { var reply = await bridge.Send(new() { ["action"] = "status" }); if (Current(token)) ApplyReply(reply); }
            catch { if (Current(token)) Failed(); }
            finally { polling = false; }
        };
        poll.Start();
        FormClosing += async (_, e) =>
        {
            if (closed) return;
            e.Cancel = true;
            await Shutdown();
            if (!IsDisposed) Close();
        };
    }

    private void Note(string text) => content.Controls.Add(new Label { Text = text, AutoSize = true, MaximumSize = new Size(540, 0) });
    private void Field(string label, TextBox field)
    {
        Note(label); field.Width = 520; field.AccessibleName = label; content.Controls.Add(field);
    }
    private void ActionButton(string text, Func<Task> action)
    {
        var button = new Button { Text = text, AccessibleName = text, AutoSize = true };
        button.Click += async (_, _) => { if (busy || closing) return; try { await action(); } catch { if (!closing) Failed(); } };
        actions.Add(button); content.Controls.Add(button);
    }
    private bool Current(int token) => !closing && !IsDisposed && token == generation;
    private void EnableActions() { foreach (var button in actions) button.Enabled = !busy && !closing; }
    private void Failed() => status.Text = "Setup could not be verified. Close this window before retrying. Saved state was retained.";
    private static int ReadPort(TextBox field) => int.TryParse(field.Text, out var value) && value is >= 1024 and <= 65535 ? value : throw new ArgumentException("Invalid port.");

    private async Task Send(JsonObject command, bool identity = false)
    {
        if (busy || closing) return;
        busy = true; var token = ++generation; EnableActions();
        try
        {
            if (identity)
            {
                var ready = await bridge.Send(new() { ["action"] = "identity-status" });
                if (!Current(token)) return;
                if (ready.Status == "identity-required" && consent.Checked)
                    ready = await bridge.PrepareIdentity(true);
                if (!Current(token)) return;
                if (ready.Status != "identity-ready")
                {
                    status.Text = ready.Status == "identity-required" ? "Review and accept the identity storage policy first." : "Identity recovery is required. Existing files will not be replaced.";
                    return;
                }
            }
            var reply = await bridge.Send(command);
            if (Current(token)) ApplyReply(reply);
        }
        catch { if (Current(token)) Failed(); }
        finally { if (Current(token)) { busy = false; EnableActions(); } }
    }
    private Task Host()
    {
        if (hosting || joining) return Task.CompletedTask;
        _ = ReadPort(port);
        return Send(new() { ["action"] = "host-start", ["address"] = address.Text.Trim(), ["port"] = 0 }, true);
    }
    private Task ConfirmHost()
    {
        if (!hosting || pendingFingerprint is null) { status.Text = "Wait for the other device to claim this invitation."; return Task.CompletedTask; }
        return Send(new() { ["action"] = "host-confirm", ["peerCertificateSha256"] = pendingFingerprint,
            ["localEndpoint"] = new JsonObject { ["kind"] = "tls", ["address"] = address.Text.Trim(), ["port"] = ReadPort(port) },
            ["peerEndpoint"] = new JsonObject { ["kind"] = "tls", ["address"] = peerAddress.Text.Trim(), ["port"] = ReadPort(peerPort) },
            ["includeUbuntu"] = ubuntu.Checked });
    }
    private Task Join()
    {
        if (hosting || joining) return Task.CompletedTask;
        return Send(new() { ["action"] = "join-claim", ["invitation"] = invitation.Text.Trim() }, true);
    }
    private void ApplyReply(TlsSetupReply reply)
    {
        switch (reply.Status)
        {
            case "hosting": hosting = true; invitation.Text = reply.Invitation; status.Text = "Share the invitation privately with your Mac."; break;
            case "confirming": pendingFingerprint = reply.PeerCertificateSha256; status.Text = "A device claimed the invitation. Confirm only if you are pairing that device now."; break;
            case "awaiting-confirmation": joining = true; status.Text = "Confirm on the host, then choose Finish joining."; break;
            case "configuration-ready": pendingFingerprint = null; status.Text = "Configuration saved. Waiting for the other device to finish joining."; break;
            case "local-ready": status.Text = "Saved locally. Acknowledgement was not verified. Retry Finish joining."; break;
            case "acknowledged": pendingFingerprint = null; status.Text = "Setup acknowledged. Live data sync and source availability are not yet verified."; break;
            case "cancelled": hosting = joining = false; pendingFingerprint = null; invitation.Clear(); status.Text = "Setup cancelled. Saved state was retained."; break;
            case "inactive": status.Text = "Invitation inactive. Close and reopen setup to retry."; break;
            case "unavailable": status.Text = "Setup unavailable. Check addresses, source scope and existing pairing state."; break;
        }
        fingerprint.Text = pendingFingerprint is null ? "" : "Peer certificate: " + pendingFingerprint;
    }

    internal Task Shutdown() => shutdown ??= Stop();

    internal static void SelfTest(string node)
    {
        var runtime = Path.Combine(Path.GetTempPath(), "observatory-pairing-window-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(runtime);
        Exception? failure = null;
        try
        {
            using var collector = new Collector(runtime);
            collector.Configure(null, activity: false, codex: false);
            var release = collector.BeginDirectPairing();
            var bridge = new TlsSetupProcess(runtime, node, Path.Combine(AppContext.BaseDirectory, "Collector", "scripts", "peer-tls-control.mjs"));
            using var window = new DirectPairingWindow(bridge, release) { Opacity = 0, ShowInTaskbar = false };
            window.Shown += async (_, _) =>
            {
                try
                {
                    if (!collector.Busy || window.consent.Checked || window.hosting || window.joining)
                        throw new Exception("Opening setup changed consent or failed to reserve collection.");
                    await window.Send(new() { ["action"] = "host-start", ["address"] = "192.168.1.2", ["port"] = 0 }, true);
                    if (window.hosting || Directory.Exists(Path.Combine(runtime, "private-device-identity")))
                        throw new Exception("Setup created an identity without consent.");
                    await window.Shutdown(); await window.Shutdown();
                    if (!bridge.ExitVerified || collector.Busy) throw new Exception("Window cleanup did not stop its helper and release collection.");
                }
                catch (Exception error) { failure = error; }
                finally { await window.Shutdown(); window.Close(); }
            };
            Application.Run(window);
            if (failure is not null) throw failure;
        }
        finally { Directory.Delete(runtime, true); }
        Console.WriteLine("Native pairing window consent rejection, helper cleanup and collector release passed.");
    }

    private async Task Stop()
    {
        closing = true; generation++; poll.Stop(); poll.Dispose(); EnableActions();
        try { await bridge.DisposeAsync(); }
        catch { /* Preserve the pause when helper exit cannot be verified. */ }
        finally { release(bridge.ExitVerified); closed = true; }
    }
}

File: native/DirectPairingWindow.swift
import AppKit
import SwiftUI

protocol DirectPairingBridge: AnyObject {
    func send(_ command: [String: Any], completion: @escaping TLSSetupProcess.Completion)
    func prepareIdentity(storageConsent: Bool, completion: @escaping TLSSetupProcess.Completion)
    func close()
}
extension TLSSetupProcess: DirectPairingBridge { }

@MainActor
final class DirectPairingModel: ObservableObject {
    @Published var address = ""
    @Published var peerAddress = ""
    @Published var port = "43128"
    @Published var peerPort = "43128"
    @Published var includeUbuntu = false
    @Published var storageConsent = false
    @Published var invitation = ""
    @Published private(set) var message = "Choose Host on one device and Join on the other. Use a local network or trusted VPN."
    @Published private(set) var peerFingerprint: String?
    @Published private(set) var busy = false
    @Published private(set) var hosting = false
    @Published private(set) var joining = false
    private let bridge: DirectPairingBridge
    private var timer: Timer?
    private var closed = false
    private var polling = false
    private var generation = 0

    init(bridge: DirectPairingBridge, polling: Bool = true) {
        self.bridge = bridge
        if polling {
            timer = Timer.scheduledTimer(withTimeInterval: 2, repeats: true) { [weak self] _ in
                Task { @MainActor [weak self] in self?.poll() }
            }
        }
    }

    func shutdown() {
        guard !closed else { return }
        closed = true; timer?.invalidate(); timer = nil; bridge.close()
    }

    private func handle(_ result: Result<TLSSetupReply, Error>) -> TLSSetupReply? {
        guard !closed else { return nil }
        guard case let .success(reply) = result else {
            message = "Setup could not be verified. Saved data was preserved. Close this window before retrying."
            return nil
        }
        switch reply.status {
        case "identity-recovery-required": message = "Identity recovery is required. Existing pairing or identity files will not be replaced."
        case "hosting": hosting = true; invitation = reply.invitation ?? ""; message = "Share this invitation privately with the other device. It expires within ten minutes."
        case "confirming": message = "A device claimed the invitation. Confirm only if you are pairing that device now."
        case "awaiting-confirmation": joining = true; message = "Confirm this device on the host, then choose Finish joining."
        case "configuration-ready": peerFingerprint = nil; message = "Configuration saved. Waiting for the other device to finish joining."
        case "local-ready": message = "Saved locally, but acknowledgement was not verified. Retry Finish joining with the same invitation."
        case "acknowledged": peerFingerprint = nil; message = "Pairing setup acknowledged. This does not verify live data sync or source availability."
        case "cancelled": hosting = false; joining = false; invitation = ""; peerFingerprint = nil; message = "Setup cancelled. Saved state was retained."
        case "inactive": message = "The invitation is inactive. Close and reopen setup to retry."
        case "unavailable": message = "Setup unavailable. Check the addresses, source scope and existing pairing state. No saved state was replaced."
        default: break
        }
        if reply.status == "confirming" { peerFingerprint = reply.peerCertificateSha256 }
        return reply
    }

    private func perform(_ command: [String: Any]) {
        guard !busy, !closed else { return }
        busy = true; generation += 1; let token = generation
        bridge.send(command) { [weak self] result in
            Task { @MainActor in guard let self, !self.closed, self.generation == token else { return }; self.busy = false; _ = self.handle(result) }
        }
    }

    private func withIdentity(_ action: @escaping @MainActor () -> Void) {
        guard !busy, !closed else { return }
        busy = true; generation += 1; let token = generation
        bridge.send(["action": "identity-status"]) { [weak self] result in
            Task { @MainActor in
                guard let self, !self.closed, self.generation == token else { return }
                guard let reply = self.handle(result) else { self.busy = false; return }
                if reply.status == "identity-ready" { self.busy = false; action(); return }
                guard reply.status == "identity-required", self.storageConsent else {
                    self.busy = false
                    if reply.status == "identity-required" { self.message = "Review and accept the identity storage policy before creating an identity." }
                    return
                }
                self.bridge.prepareIdentity(storageConsent: true) { [weak self] result in
                    Task { @MainActor in
                        guard let self, !self.closed, self.generation == token else { return }; self.busy = false
                        if self.handle(result)?.status == "identity-ready" { action() }
                    }
                }
            }
        }
    }

    func host() {
        guard !hosting, !joining, let value = Int(port), (1024...65535).contains(value), !address.isEmpty else {
            message = "Enter this device's numeric private-network address and a port from 1024 to 65535."; return
        }
        withIdentity { [weak self] in guard let self else { return }; self.perform(["action": "host-start", "address": self.address, "port": 0]) }
    }

    func confirmHost() {
        guard hosting, let fingerprint = peerFingerprint, let local = Int(port), let remote = Int(peerPort),
              (1024...65535).contains(local), (1024...65535).contains(remote), !peerAddress.isEmpty else {
            message = "Wait for a device claim and enter the other device's private-network address and sync port."; return
        }
        perform(["action": "host-confirm", "peerCertificateSha256": fingerprint,
                 "localEndpoint": ["kind": "tls", "address": address, "port": local],
                 "peerEndpoint": ["kind": "tls", "address": peerAddress, "port": remote], "includeUbuntu": includeUbuntu])
    }

    func join() {
        guard !hosting, !joining, invitation.utf8.count <= 2048, invitation.hasPrefix("observatory-pair:v1:") else {
            message = "Paste a complete invitation from the other device."; return
        }
        withIdentity { [weak self] in guard let self else { return }; self.perform(["action": "join-claim", "invitation": self.invitation]) }
    }

    func finishJoining() { if joining { perform(["action": "join-confirm", "includeUbuntu": false]) } }

    func cancel() {
        guard !closed else { return }
        generation += 1; let token = generation
        bridge.send(["action": "cancel"]) { [weak self] result in
            Task { @MainActor in guard let self, !self.closed, self.generation == token else { return }; self.busy = false; _ = self.handle(result) }
        }
    }

    func poll() {
        guard !closed, !polling, !busy, hosting else { return }
        polling = true; let token = generation
        bridge.send(["action": "status"]) { [weak self] result in
            Task { @MainActor in
                guard let self, !self.closed else { return }; self.polling = false
                if self.generation == token { _ = self.handle(result) }
            }
        }
    }

    func copyInvitation() {
        guard hosting, !invitation.isEmpty else { return }
        NSPasteboard.general.clearContents(); NSPasteboard.general.setString(invitation, forType: .string)
    }
}

private struct DirectPairingView: View {
    @ObservedObject var model: DirectPairingModel
    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            Text("Direct device pairing").font(.title2.weight(.semibold))
            Text("Pairing is separate from provider sign-in. This setup does not enable allowance sharing or prove that data sync is running.")
                .font(.callout).foregroundStyle(.secondary)
            Toggle("Allow a device identity stored in a permission-restricted file", isOn: $model.storageConsent)
            Text("The key is not stored in Keychain or encrypted separately by Observatory. A valid existing identity is reused without creating another.")
                .font(.caption).foregroundStyle(.secondary)
            GroupBox("Host on this Mac") {
                VStack(alignment: .leading, spacing: 10) {
                    HStack { TextField("This Mac's private IP", text: $model.address); TextField("Sync port", text: $model.port).frame(width: 90) }
                    HStack { TextField("Windows private IP", text: $model.peerAddress); TextField("Peer sync port", text: $model.peerPort).frame(width: 90) }
                    Toggle("Include Ubuntu Codex records on Windows", isOn: $model.includeUbuntu)
                        .disabled(model.joining)
                    Text("This option applies when hosting on this Mac. When this Mac joins an invitation, choose Ubuntu scope on the Windows host.")
                        .font(.caption).foregroundStyle(.secondary)
                    HStack {
                        Button("Create invitation", action: model.host).disabled(model.hosting || model.joining)
                        Button("Confirm this device", action: model.confirmHost).disabled(model.peerFingerprint == nil)
                    }
                }.padding(6)
            }.disabled(model.busy)
            GroupBox("Invitation") {
                VStack(alignment: .leading, spacing: 8) {
                    TextField("Paste an invitation, or create one above", text: $model.invitation).textFieldStyle(.roundedBorder)
                    HStack {
                        Button("Copy invitation", action: model.copyInvitation).disabled(!model.hosting)
                        Button("Join", action: model.join).disabled(model.hosting || model.joining)
                        Button("Finish joining", action: model.finishJoining).disabled(!model.joining)
                    }
                    Text("Share privately. Clipboard managers may retain copied invitations.").font(.caption).foregroundStyle(.secondary)
                }.padding(6)
            }.disabled(model.busy)
            if let fingerprint = model.peerFingerprint {
                Text("Claimed device: \(fingerprint)").font(.caption.monospaced()).textSelection(.enabled)
            }
            Text(model.message).font(.callout).fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
            Button("Cancel setup", action: model.cancel)
        }.padding(22).frame(minWidth: 560, minHeight: 540)
    }
}

@MainActor
final class DirectPairingWindowController: NSWindowController, NSWindowDelegate {
    private let model: DirectPairingModel
    private let onClose: () -> Void
    init(bridge: DirectPairingBridge, onClose: @escaping () -> Void) {
        model = DirectPairingModel(bridge: bridge); self.onClose = onClose
        let window = NSWindow(contentRect: NSRect(x: 0, y: 0, width: 620, height: 600),
                              styleMask: [.titled, .closable, .resizable], backing: .buffered, defer: false)
        super.init(window: window)
        window.title = "Observatory pairing"; window.isReleasedWhenClosed = false
        window.minSize = NSSize(width: 600, height: 600)
        window.contentView = NSHostingView(rootView: DirectPairingView(model: model))
        window.delegate = self; window.center()
    }
    required init?(coder: NSCoder) { fatalError("Not supported") }
    func windowWillClose(_ notification: Notification) { model.shutdown(); onClose() }
}
